### ISSUE TITLE
Terraform provider not syncing with GitHub

### ISSUE THREAD

**Issue by nikwotton** [integrations/terraform-provider-github#1098]
Hello! :wave:

I am having problems with Terraform's GitHub provider not syncing with GitHub. I do not know for sure whether it is an issue with Terraform or the provider, so I am starting here. If this should be reported to Terraform instead please let me know. 

Below I have outlined one such example of syncing failing. Note that I am able to reproduce this with any of several different configuration changes, but all follow the same basic steps with the only difference being what configuration step is made. The basic idea is that if a configuration changes and then fails to apply with terraform, in this case due to a permissions mis-match, the tfstate is changed ignoring the failure and the tfstate never re-syncs with GitHub to check if the tfstate is still correct. 

While the permissions thing is bad, what I am mainly worried about is any kind of failure leading to this weird state where terraform says it is in sync when in fact it is not, and there is no way to know that there is an issue other than checking setting by setting, repo by repo to determine whether the state is correct. This is not an acceptable position to be in as it removes the benefit of automation and actually worsens the situation. Per [Terraform's documentation](https://www.terraform.io/cli/commands/plan#planning-options) a re-sync should occur automatically and by default when running `terraform plan` which is not seen in the below steps. 

### Terraform Version
```
$ terraform -v
Terraform v1.1.7
on darwin_amd64
+ provider registry.terraform.io/integrations/github v4.23.0
```

### Affected Resource(s)
- github_repository
- Likely others, this resource is what's currently in question though

### Terraform Configuration Files
```hcl
terraform {
  // confirmed latest
  required_version = "1.1.7"
  required_providers {
    github = {
      source  = "integrations/github"
      // confirmed latest
      version = "4.23.0"
    }
  }
}
variable "GITHUB_TOKEN" {
  type = string
}

provider "github" {
  token = var.GITHUB_TOKEN
  // here <REDACTED> is the name of a GitHub Organization, not a user
  owner = "<REDACTED>"
}
resource "github_repository" "archivable-repo" {
  // archivable-repo is a real repository under the <REDACTED> GitHub Organization
  name          = "archivable-repo"
  // allow_squash_merge is the example used for the purposes of this GitHub Issue. Several others have been tested as well though
  allow_squash_merge = true // Depending on step
  // These are only set because the defaults don't match the repository. They are unrelated to reproducing this issue
  has_downloads = true
  has_issues = true
  has_projects = true
  has_wiki = true
}
```

### Debug Output
These Gists includes the full outputs of several steps from the `Steps to Reproduce` below. Each section is called out for which step it is a part of. Note that not all steps below contain commands to run so not all are represented in this log. The only redaction in these logs are the GitHub Organization, so anywhere you see `<REDACTED>` just assume that means the name of the org. The only difference between the gists is the inclusion of the TRACE log level - I wasn't sure whether that was/is wanted or just debug. 
With TRACE: https://gist.github.com/nikwotton/fb6f89db9e3137337901bc729658dcb5
With DEBUG: https://gist.github.com/nikwotton/968b98539e546a1f5f1bddc4a7f4ba3a

### Expected Behavior
Given that what's reported by GitHub is different from what is locally in the tfstate, the tfstate should be updated when running `terraform plan` and/or `terraform apply` and the plans should reflect that a change is being requested. 

### Actual Behavior
The tfstate is not updated, terraform claims no changes will be made, and the plan says "No changes. Your infrastructure matches the configuration." which is not in fact the case. 

### Steps to Reproduce
1. Setup your configuration to match the actual settings of a repository you managed on GitHub. For example, copy the above configuration to a `.tf` file and create a repository in your Organization named archivable-repo
2. Run `terraform init`
3. Import the actual repository using `terraform import github_repository.archivable-repo archivable-repo`
4. Run `terraform plan` to ensure that your configuration matches what is on GitHub, or change the configuration until it matches. 
5. Change a flag that the token you are using does not have permission to change (here `allow_squash_merge` from true to false)
6. Run `terraform apply` and tell terraform to make the change. Note that this will fail with a 404 and no change will be made on GitHub. 
7. Grant the token you are using access to admin the repository, or setup a separate access token locally that does have admin access
8. Run `terraform plan` and see again the "No changes. Your infrastructure matches the configuration.", even though you can verify that the configuration is now different from what is on GitHub. 
9. Run `terraform apply` and again see that no changes are to be made, even though the configuration does not match what is on GitHub. 

Note that the situation can be resolved easily by dropping the repository from state and re-importing it, but that is not the point here. If no error is reported, there is no way to know that a repo is in need of being fixed, and regularly checking every setting in every repo to see if it matches the configuration is far worse than simply not using terraform.

**Comment by nikwotton** [comment]
I just thought to check if it was something to do with caching due to using the same access tokens and changing permissions elsewhere. I just generated a new token with admin access to the repo to test this theory and terraform now does notice a difference! The plan now sees the change that should be made as well as an etag change, which I assume relates to the new key (I don't know much about etags). 

This tells me that it is definitely a caching issue. I do not know from that whether the issue is specific to security changes or if any network error when running `terraform apply` could lead to this caching issue, and if the latter is true then I do not know whether swapping the key would fix other scenarios. More investigation is needed

**Comment by Horgix** [comment]
I feel like this is a duplicate of #1019 (which also seems to have more details), isn't it?

Asking because I stumble upon both issues when investigating that bug on my side :slightly_smiling_face:

**Comment by kfcampbell** [comment]
It does indeed appear to be a duplicate of #1019. Let's close this issue and continue the conversation there.
